### PR TITLE
Java errors - Resolves #305

### LINKED ISSUE #305: [Entry Request] Java: Errors
## Description

Write a high-quality "Errors" entry in Java. 

It should include some of the most common errors, exceptions, and even some debugging techniques.

Java topic: https://www.codecademy.com/resources/docs/java

## Type of Bug Request

- Typo and/or bug within one or more existing entries.
- Unclear explanation of term(s).
- Out-of-date information (new release, deprecated package, etc).
- Updating the documentation

## Type of Entry Request

- One or more new entries/terms requested.
- Recommend change(s) to documentation (such as existing text or new text)

## Checklist

- [x] The issue is appropriately titled.
- [x] Description is clear and includes supporting text, files, etc.
- [x] A similar issue hasn't already been created.
- [x] No more than 3 entries requested.

## For Maintainers

- [x] Labels are added
- [ ] Assignee is added

### PR DESCRIPTION
**Pull request by SSwiniarski**
### Description

Adding "Errors" to docs/content/java/concepts/

### Type of Change

- Adding a new entry

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.

### CODE REVIEW

File: content/java/concepts/errors/errors.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+```
+
+## Example Errors
+
+- `NoClassDefFoundError` - Java cannot find the file with the class containing the `main()` method.
+- `NoSuchMethodFoundError` - Java cannot find a method being called in the program.
+- `ClassFormatError` - Java cannot read a class file or cannot interpret it.
+- `IncompatibleClassChangeError` - Code has tried to change a base class after a child class has been initalized.
+- `NoSuchFieldError` - Code has tried to access a field in an object and the field no longer exists.
```
Comment (yangc95) [inline content/java/concepts/errors/errors.md:51]:
```suggestion
- The `NoClassDefFoundError` occurs when the file with the class containing the `main()` method cannot be found.
- The `NoSuchMethodFoundError` occurs when a method is called in the program but cannot be found.
- The `ClassFormatError` occurs when a class file cannot be read or interpreted.
- The `IncompatibleClassChangeError` occurs when a base class is altered after a child class has already been initialized.
- The `NoSuchFieldError` occurs when an attempt to access a field in an object is made but the field no longer exists.
```

File: content/java/concepts/errors/errors.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+- `ArrayIndexOutOfBoundsException` - Code has tried to access an array with an index outside the established values.
+- `StringIndexOutOfBoundsException` - Code has tried to access part of a string past the boundaries of the string.
+- `NullPointerException` - Code is trying to access an object reference that does not have an instance assigned to it.
+- `ArrayStoreException` - Code has violated the rules for casting elements into arrays.
+- `ClassCastException` - Code is trying to cast an object to the wrong class.
+- `FileNotFoundException`- Java cannot find a file with the specified path.
+- `AccessControlException` - Java has been denied access to a system resource (file, network, etc.).
+- `EOFException` - Code has unexpectedly reached the end of a file or stream during input.
+- `NumberFormatException` - Code has tried to convert a string to a numeric type and the string is not a number.
```
Comment (yangc95) [inline content/java/concepts/errors/errors.md:63]:
```suggestion
- The `ArrayIndexOutOfBoundsException` occurs when an attempt to access an array is made with an index outside the established values.
- The `StringIndexOutOfBoundsException` occurs when an attempt is made to access an out-of-range part of a string.
- The `NullPointerException` occurs when an attempt is made to access an object reference that does not have an instance assigned to it.
- The `ArrayStoreException` occurs when the rules for casting elements into arrays are violated.
- The `ClassCastException` occurs when an object is cast to the wrong class.
- The `FileNotFoundException`occurs when a file with the specified path cannot be found.
- The `AccessControlException` occurs when access to a system resource (file, network, etc.) has been denied.
- The `EOFException` occurs when the end of a file or stream during input is unexpectedly reached.
- The `NumberFormatException` occurs when an attempt is made to convert a string to a numeric type and the string does not contain a number.
```

**Review by yangc95 — COMMENTED** [review]
@SSwiniarski I went ahead and did a quick first review for this entry 📝

**Comment by SSwiniarski** [comment]
@yangc95, I adopted all your suggestions. We're ready for review #2.

**Review by yangc95 — APPROVED** [review]

**Comment by Dusch4593** [comment]
Ready to to do review 2️⃣!

File: content/java/concepts/errors/errors.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+---
+Title: 'Errors'
+Description: 'In Java there are two subclasses of the java.lang.Throwable class, Errors and Exceptions.'
+Subjects:
+  - 'Computer Science'
+  - 'Code Foundations'
+Tags:
+  - 'Errors'
+  - 'Error Handling'
+  - 'Debugging'
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:10]:
```suggestion
  - 'Errors'
  - 'Error Handling'
  - 'Exceptions'
  - 'Debugging'
```

File: content/java/concepts/errors/errors.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+CatalogContent:
+  - 'learn-java'
+  - 'paths/computer-science'
+---
+
+In Java, there are two subclasses of the `java.lang.Throwable` class:
+
+- The `Error` class shows illegal operations in the code that cause an abnormal state in the program. An error is a serious abnormality in the program that the programmer should not try and `catch` during runtime.
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:18]:
```suggestion
- The `Error` class represents illegal operations that cause serious abnormalities in the program and are not recommended to `catch`.
```

File: content/java/concepts/errors/errors.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+- The `Exception` class refers to an abnormal or unexpected event that disrupts the flow of the program. Exceptions are those conditions that a reasonable programmer might try to `catch`.
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:19]:
```suggestion
- The `Exception` class refers to abnormal, unexpected events that disrupt the flow of the program but can be reasonably handled by a `catch`-statement.
```

File: content/java/concepts/errors/errors.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+## Try Catch Syntax
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:21]:
```suggestion
## Syntax
```

File: content/java/concepts/errors/errors.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+A `try...catch` block is a means for a programmer to encapsulate a block of code and "catch" a potential `java.lang.Throwable` class being thrown before it halts execution of the program. Note that programs should not `try...catch` `Errors`, only `Exceptions`.
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:23]:
```suggestion
A `try...catch` block is a means for a programmer to encapsulate a block of code and "catch" a potentially thrown `Exception` (but never an `Error`) before it halts execution of the program. 
```

File: content/java/concepts/errors/errors.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  public static void main(String[] args) {
+    int a = 27, b = 0;
+    try {
+      // The code wrapped by the try block will run until an Exception is thrown
+      System.out.println("I'm executed first!");
+      int c = a / b;
+      System.out.println("I'm never executed!");
+    } catch (ArithmeticException e) {
+      // If the above Exception is thrown, code will resume execution here
+      // This block is not run if no Exception is thrown
```
Comment (Dusch4593) [inline content/java/concepts/errors/errors.md:36]:
```suggestion
      // If the ArithmeticException is thrown above, code execution will resume here
```

**Comment by SSwiniarski** [comment]
@Dusch4593, I implemented your suggestions and we're ready to go.

**Comment by Dusch4593** [comment]
@SSwiniarski Cool cool! Before we merge, I made a couple of changes in 351f0e2fb4b11f183b9ef0fea5788571da796d5a including:

- Changed header from "## Syntax" to "## In Practice" since we're using real Java code that can be copy/pasted and applied in practice.
- Remove most of the comments in the code snippet and incorporated in a brief paragraph beneath it.
- Other small, typographical edits.

How do things look on everyone's end?

cc: @yangc95

**Review by Dusch4593 — APPROVED** [review]

**Comment by yangc95** [comment]
@Dusch4593 Interestingly, I don't see any codebytes for Java. Does Docs product support Java codebytes?

**Comment by Dusch4593** [comment]
> @Dusch4593 Interestingly, I don't see any codebytes for Java. Does Docs product support Java codebytes?

Good point! 😌 That's a bummer, tho. I wish Java was supported.

**Comment by Dusch4593** [comment]
Merged! Thanks a lot, @SSwiniarski! The entry is now live: https://www.codecademy.com/resources/docs/java/errors

**Comment by christian-dinh** [comment]
Sorry @yangc95 I was OOO! Codebytes does support Java for future reference.

**Comment by christian-dinh** [comment]
wait never mind I guess it doesn't (I could have sworn it did). If you're curious what languages are supported you can visit https://www.codecademy.com/codebyte-editor and look through the dropdown menu

**Comment by christian-dinh** [comment]
no plans for the near future unfortunately, at least as far as I'm aware.

**Comment by Dusch4593** [comment]
No worries! Was just curious 🧐

**Pull request merged**